clients/src/main/java/org/apache/kafka/common/requests/ProduceRequest.java

```diff
@@ -62,7 +62,7 @@ public Builder(byte magic,
                        int timeout,
                        Map<TopicPartition, MemoryRecords> partitionRecords,
                        String transactionalId) {
-            super(ApiKeys.PRODUCE, (short) (magic == RecordBatch.MAGIC_VALUE_V2 ? 3 : 2));
+            super(ApiKeys.PRODUCE, (short) (magic == RecordBatch.MAGIC_VALUE_V2 ? ApiKeys.PRODUCE.latestVersion() : 2));
```

Comment:
I think the `3` being hard coded here was a bug. Once the desired version is set, that is always the protocol version used. I think we should always use the `latestVersion` if we are using v2 of the message format.